text/2019-02-26-tikv-client-command-line.md

```diff
@@ -0,0 +1,208 @@
+
+Users will interact with the tool through the command line. It will support both
+a one-off mode or a REPL (Read-eval-print-loop) style mode (See alternatives).
+
+Example of the one-off raw mode:
+
+```bash
+$ tikv-cli raw set "cncf" "linux foundation"
+Finished in 0.001s.
+$ tikv-cli raw set "tikv" "rust"
+Finished in 0.001s.
+$ tikv-cli raw get "cncf"
+"linux foundation"
+Finished in 0.002s.
+$ tikv-cli raw scan "b"..
+[
+    { key: "cncf", value: "linux foundation" },
+    { key: "tikv", value: "rust" },
+]
+Finished in 0.003s.
+```
+
+Example of REPL raw mode:
+
+```bash
+$ tikv-cli raw
+> set "cncf" "linux foundation"
+Finished in 0.001s.
+> set "tikv" "rust"
+Finished in 0.001s.
+> get "cncf"
+"linux foundation"
+Finished in 0.001s.
+> scan "b"..
+[
+    { key: "cncf", value: "linux foundation" },
+    { key: "tikv", value: "rust" },
+]
+Finished in 0.003s.
+```
+
+REPL transactional mode, notice how the `>>` communicates that the user is in a
+transaction. Also notice how the transaction timestamp is outputted in the
+request complete messages.
+
+```bash
+$ tikv-cli txn
+> begin
+Finished in 0.000s. (txn ts: 1551216224838)
+>>set "cncf" "linux foundation"
+Finished in 0.001s. (txn ts: 1551216224838)
+>> set "tikv" "rust"
+Finished in 0.001s. (txn ts: 1551216224838)
+>> get "cncf"
+"linux foundation"
+Finished in 0.001s. (txn ts: 1551216224838)
+>> scan "b"..
+[
+    { key: "cncf", value: "linux foundation" },
+    { key: "tikv", value: "rust" },
+]
+Finished in 0.003s. (txn ts: 1551216224838)
+> commit
+Finished in 0.001s. (txn ts: 1551216224838)
+```
+
+One-off transactional mode is a bit different, as we need some way to maintain
+timestamp across commands. The user may choose how to do this, the reccomended
+way is via ENV variables, but they must do this. (See alternatives)
+
+```bash
+$ TS=$(tikv-cli txn begin)
+$ tikv-cli txn ${TS} set "cncf" "linux foundation"
+Finished in 0.001s. (txn ts: 1551216224838)
+$ tikv-cli txn ${TS} set "tikv" "rust"
+Finished in 0.001s. (txn ts: 1551216224838)
+$ tikv-cli txn ${TS} get "cncf"
+"linux foundation"
+Finished in 0.002s. (txn ts: 1551216224838)
+$ tikv-cli txn ${TS} scan "b"..
+[
+    { key: "cncf", value: "linux foundation" },
+    { key: "tikv", value: "rust" },
+]
+Finished in 0.003s. (txn ts: 1551216224838)
+$ tikv-cli txn ${TS} commit
+Finished in 0.004s. (txn ts: 1551216224838)
+```
+
+## Drawbacks
+
+We may not wish to need to maintain this tool, or we may wish for the community
+to maintain it.
+
+## Alternatives
+
+* *(Installation/Repo)* We could use the `tikv-client` repository and provide
+  either an opt-out or an opt-in binary from `src/bin/tikv-cli.rs`. If we chose
+  an opt-in functionality here users mgiht end up with an awkard installation
+  call.
+* (*Use `tikv-cli`*), we could use `tikv-cli` for this, but it creates as fuzzy
+  distinction between management tools and clients.
+* (*No REPL, or only REPL*). We may chose to not have a REPL mode, or we may chose

```

Comment:
I rarely use one off mode of mysql client. But I think running commands from file is very useful.

Reply:
Is there an advantage to just piping to the REPL?

Reply:
Hm, it actually doesn't make much difference. It's just something that I would expect to have as it is in mysql and hive client.

Reply:
What do you think @siddontang ?

Reply:
I think it is good to have both, and I don't think it adds too much complexity (as long as we can remove the need for raw and txn modes as described above).